### ISSUE TITLE
Add support for local version identifiers

### ISSUE THREAD

**Issue by Gideon357** [pypa/warehouse#7989]
**tl;dr**: Add support for Local Version Identifiers (i.e. `pip install torch+cpu` or `poetry add package+windows`)

**What's the problem this feature will solve?**
PyPi should support local version identifiers, so you could have something like `torch+cpu` in the same package as `torch`, and directly `pip install torch+cpu` or `poetry add torch+cpu` instead of `pip3 install torch==1.2.0+cpu torchvision==0.4.0+cpu -f https://download.pytorch.org/whl/torch_stable.html`. I also see how this could work for packages that use different wheels for different OSs. I found a number of places this could be useful, including PyTorch/pytorch#26340. Additionally, [PEP #440](https://www.python.org/dev/peps/pep-0440/#local-version-identifiers) calls for use of local version identifiers, so it makes sense for PyPi to support them.  Overall, this would simplify installation for a number of packages, and improve PyPi, especially in regard to how it works with package managers like [Poetry](github.com/python-poetry/poetry)

**Describe the solution you'd like**
Add support for Local Version Identifiers (i.e. `pip install torch+cpu` or `poetry add package+windows`)

**Comment by pradyunsg** [comment]
From PEP 440:

> Local version identifiers SHOULD NOT be used when publishing upstream projects to a public index server, but MAY be used to identify private builds created directly from the project source.

---

Are you looking for more granular [environment markers](https://www.python.org/dev/peps/pep-0508/#environment-markers) or something to determine dependencies based on the target system's capabilities?

**Comment by di** [comment]
Thanks for feature request.

Closing since PEP 440 explicitly states that local versions should not be used on upstream indices like PyPI.
